akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolGateway.scala

```diff
@@ -0,0 +1,139 @@
+ */
+
+package akka.http.engine.client
+
+import scala.annotation.tailrec
+import scala.concurrent.{ Future, Promise }
+import scala.concurrent.duration.{ Duration, FiniteDuration }
+import scala.util.{ Failure, Try }
+import akka.actor.{ ActorLogging, Cancellable }
+import akka.stream.FlowMaterializer
+import akka.stream.actor.{ ActorPublisher, ActorSubscriber, ZeroRequestStrategy }
+import akka.stream.actor.ActorPublisherMessage._
+import akka.stream.actor.ActorSubscriberMessage._
+import akka.stream.impl.ResizableMultiReaderRingBuffer
+import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.http.model._
+
+private object HostConnectionPoolGateway {
+  import ConnectionPool.{ PoolRequest, BeginShutdown }
+
+  case class RequestContext(request: HttpRequest, responsePromise: Promise[HttpResponse], retriesLeft: Int) {
+    require(retriesLeft >= 0)
+  }
+  case class ResponseContext(rc: RequestContext, response: Try[HttpResponse])
+
+  private case object ShutdownNow
+
+  class PoolGateway(poolFlow: Flow[RequestContext, ResponseContext, Any],
+                    hcps: HostConnectionPoolSetup,
+                    whenShuttingDown: Promise[Future[Unit]])(implicit fm: FlowMaterializer)
+    extends ActorSubscriber with ActorPublisher[RequestContext] with ActorLogging
+    with ResizableMultiReaderRingBuffer.Cursors with ResizableMultiReaderRingBuffer.Cursor {
+    import context.dispatcher
+
+    val cursors = this :: Nil // ResizableMultiReaderRingBuffer.Cursors
+    var cursor: Int = _ // ResizableMultiReaderRingBuffer.Cursor
+    private[this] val inputBuffer = {
+      val maxOpenRequests = hcps.setup.settings.maxOpenRequests
+      new ResizableMultiReaderRingBuffer[PoolRequest](maxOpenRequests, maxOpenRequests, this)
+    }
+    private[this] var activeIdleTimeout: Option[Cancellable] = None
+    private[this] val effectiveDefaultHeaders =
+      if (!hcps.setup.defaultHeaders.exists(_.isInstanceOf[headers.Host])) {
+        import Uri._
+        val normalizedPort = normalizePort(hcps.port, httpScheme(securedConnection = hcps.setup.encrypted))
+        headers.Host(hcps.host, normalizedPort) +: hcps.setup.defaultHeaders
+      } else hcps.setup.defaultHeaders
+
+    log.debug("(Re-)starting host connection pool to {}:{}", hcps.host, hcps.port)
+
+    // start the pool flow with this actor acting as source as well as sink
+    Source(ActorPublisher(self)).via(poolFlow).runWith(Sink(ActorSubscriber[ResponseContext](self)))
+
+    activateIdleTimeoutIfNecessary()
+
+    def requestStrategy = ZeroRequestStrategy
+
+    def receive: Receive = {
+
+      /////////////// FROM POOL DOWNSTREAM //////////////
+
+      case Request(_) ⇒ dispatchRequests() // the pool is ready to take on more requests
+
+      case Cancel ⇒
+        // somehow the pool shut down, however, we don't do anything here because we'll also see an
+        // OnComplete or OnError which we use as the sole trigger for cleaning up
+        ()
```

Comment:
It's here to keep scalariform happy. Otherwise the comment will be strangely aligned.